# Teclov investment data Analysis
Teclov is an investment company who want to invest 5M-15M in any investment sector in any country. Which country and which sector would be better, we need to find it.

The dataset for this analysis is taken from a "kaggle competition"

In [1]:
# importing required library
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
# let's import and read the data files

rounds = pd.read_csv('rounds2.csv', sep = ',', encoding='ISO-8859-1')
companies = pd.read_csv('companies.txt', sep = '\t', encoding='ISO-8859-1')

In [4]:
# let's look at the round dataset

print(rounds.head())

                     company_permalink  \
0                  /organization/-fame   
1               /ORGANIZATION/-QOUNTER   
2               /organization/-qounter   
3  /ORGANIZATION/-THE-ONE-OF-THEM-INC-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015         10000000.0  
1                  A  14-10-2014                NaN  
2                NaN  01-03-2014           700000.0  
3                  B  30-01-2014          3406878.0  
4                  A  19-03-2008      

In [6]:
# let's inspect the the structure of the dataset
print(rounds.info(), '\n')
print(rounds.shape)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 114949 entries, 0 to 114948
Data columns (total 6 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   company_permalink        114949 non-null  object 
 1   funding_round_permalink  114949 non-null  object 
 2   funding_round_type       114949 non-null  object 
 3   funding_round_code       31140 non-null   object 
 4   funded_at                114949 non-null  object 
 5   raised_amount_usd        94959 non-null   float64
dtypes: float64(1), object(5)
memory usage: 3.1+ MB
None 

(114949, 6)


in the dataset rounds.csv the funding_round_code and the raised_amount_usd have some missing values which will be taken care later.

now let's inspect the companies file

In [8]:
companies.head()

permalink                    name  \
0                  /Organization/-Fame                   #fame   
1               /Organization/-Qounter                :Qounter   
2  /Organization/-The-One-Of-Them-Inc-  (THE) ONE of THEM,Inc.   
3                /Organization/0-6-Com                 0-6.com   
4       /Organization/004-Technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating          NaN   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010

In [9]:
# let's see the structure of the companies file

companies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66368 entries, 0 to 66367
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   permalink      66368 non-null  object
 1   name           66367 non-null  object
 2   homepage_url   61310 non-null  object
 3   category_list  63220 non-null  object
 4   status         66368 non-null  object
 5   country_code   59410 non-null  object
 6   state_code     57821 non-null  object
 7   region         58338 non-null  object
 8   city           58340 non-null  object
 9   founded_at     51147 non-null  object
dtypes: object(10)
memory usage: 2.5+ MB


Here the permalink columns will be the primary key of the dataset which indicate that there will be 66368 unique companies.
& these 66368 companies must be present in round file.

In [11]:
# let's identify the unique companies or parmalinks

len(companies.permalink.unique())

66368

In [12]:
# now let's convert all the character into lower case in case there might have uppercase
companies['permalink'] = companies['permalink'].str.lower()
companies.head()

permalink                    name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :Qounter   
2  /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
3                /organization/0-6-com                 0-6.com   
4       /organization/004-technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating          NaN   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010

In [14]:
# let's look at the unique value again
len(companies.permalink.unique())

66368

This look like the permalinks are same as previous. Which mean that there will be 66368 companies in the dataset. These companies must be present in rounds file also.

In [15]:
# let's check how many companies are there in round file
len(rounds.company_permalink.unique())

90247

In [16]:
# there are so many permalinks in round file. first let's make them uniform

rounds['company_permalink'] = rounds['company_permalink'].str.lower()
rounds.head()

company_permalink  \
0                  /organization/-fame   
1               /organization/-qounter   
2               /organization/-qounter   
3  /organization/-the-one-of-them-inc-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015         10000000.0  
1                  A  14-10-2014                NaN  
2                NaN  01-03-2014           700000.0  
3                  B  30-01-2014          3406878.0  
4                  A  19-03-2008          2000000.0

In [20]:
# let's check the number of permalink again
len(rounds.company_permalink.unique())

66370

 it looks like there are 2 permalink extra than that of company file. 

Now let's have a look at the company permalink which are not in round permalink

In [21]:
rounds.loc[~rounds['company_permalink'].isin(companies['permalink']), :]

company_permalink  \
29597                             /organization/e-cãŠbica   
31863        /organization/energystone-games-灵石游戏   
45176                /organization/huizuche-com-惠ç§ÿ车   
58473              /organization/magnet-tech-磁石科技   
101036  /organization/tipcat-interactive-沙èˆÿ信息ç...   
109969             /organization/weiche-tech-喂车科技   
113839                 /organization/zengame-ç¦
游科技   

                                funding_round_permalink funding_round_type  \
29597   /funding-round/8491f74869e4fe8ba9c378394f8fbdea               seed   
31863   /funding-round/b89553f3d2279c5683ae93f45a21cfe0               seed   
45176   /funding-round/8f8a32dbeeb0f831a78702f83af78a36               seed   
58473   /funding-round/8fc91fbb32bc95e97f151dd0cb4166bf               seed   
101036  /funding-round/41005928a1439cb2d706a43cb661f60f               seed   
109969  /funding-round/f74e457f838b81fa0b29649740f186d8            venture   
113839  /funding-round/6ba28fb4f3eadf5a9c6c81bc5dde6cdf               seed   

       funding_round_code   funded_at  raised_amount_usd  
29597                 NaN  01-02-2015                NaN  
31863                 NaN  09-08-2014                NaN  
45176                 NaN  18-09-2014                NaN  
58473                 NaN  16-08-2014          1625585.0  
101036                NaN  06-09-2010                NaN  
109969                  A  06-09-2015                NaN  
113839                NaN  17-07-2010                NaN

In company permalink these characters are weired which may caused due to encoding problem. Let's check the origianl datafile to understand more

In [23]:
# let's check the origial dataset
round_original = pd.read_csv('rounds2.csv', encoding = "ISO-8859-1")
round_original.iloc[[29597, 31863, 45176, 58473, 101036, 113839], :]

company_permalink  \
29597                             /ORGANIZATION/E-CÊBICA   
31863        /ORGANIZATION/ENERGYSTONE-GAMES-ǁµÇŸ³Æ¸¸Æˆ   
45176                /organization/huizuche-com-惠ç§ÿ车   
58473              /ORGANIZATION/MAGNET-TECH-Ç£ÇŸ³Ç§‘ÆŠ€   
101036  /organization/tipcat-interactive-沙èˆÿ信息ç...   
113839                 /ORGANIZATION/ZENGAME-Ǧ
Ƹ¸Ç§‘ÆŠ€   

                                funding_round_permalink funding_round_type  \
29597   /funding-round/8491f74869e4fe8ba9c378394f8fbdea               seed   
31863   /funding-round/b89553f3d2279c5683ae93f45a21cfe0               seed   
45176   /funding-round/8f8a32dbeeb0f831a78702f83af78a36               seed   
58473   /funding-round/8fc91fbb32bc95e97f151dd0cb4166bf               seed   
101036  /funding-round/41005928a1439cb2d706a43cb661f60f               seed   
113839  /funding-round/6ba28fb4f3eadf5a9c6c81bc5dde6cdf               seed   

       funding_round_code   funded_at  raised_amount_usd  
29597                 NaN  01-02-2015                NaN  
31863                 NaN  09-08-2014                NaN  
45176                 NaN  18-09-2014                NaN  
58473                 NaN  16-08-2014          1625585.0  
101036                NaN  06-09-2010                NaN  
113839                NaN  17-07-2010                NaN

This problem is created due to encoding problem. Let's find out which type of encoding is used here

In [26]:
import chardet
rawdata = open('rounds2.csv', 'rb').read()
result = chardet.detect(rawdata)
charenc = result['encoding']
print(charenc)

Windows-1254


Apparently, pandas cannot decode "cp1254" in this case. Now let's try alternative way to decode the column

In [29]:
rounds['company_permalink'] = rounds['company_permalink'].str.encode('utf-8').str.decode('ascii', 'ignore')
rounds.loc[~rounds['company_permalink'].isin(companies['permalink']),:]

company_permalink  \
77                                  /organization/10north   
729                               /organization/51wofang-   
2670                              /organization/adslinked   
3166    /organization/aesthetic-everything-social-network   
3291                 /organization/affluent-attach-club-2   
...                                                   ...   
110545                       /organization/whodats-spaces   
113839                             /organization/zengame-   
114946                                 /organization/eron   
114947                               /organization/asys-2   
114948  /organization/novatiff-reklam-ve-tantm-hizmetl...   

                                funding_round_permalink   funding_round_type  \
77      /funding-round/b41ff7de932f8b6e5bbeed3966c0ed6a  equity_crowdfunding   
729     /funding-round/346b9180d276a74e0fbb2825e66c6f5b              venture   
2670    /funding-round/449ae54bb63c768c232955ca6911dee4                 seed   
3166    /funding-round/62593455f1a69857ed05d5734cc04132  equity_crowdfunding   
3291    /funding-round/626678bdf1654bc4df9b1b34647a4df1                 seed   
...                                                 ...                  ...   
110545  /funding-round/d5d6db3d1e6c54d71a63b3aa0c9278e6                 seed   
113839  /funding-round/6ba28fb4f3eadf5a9c6c81bc5dde6cdf                 seed   
114946  /funding-round/59f4dce44723b794f21ded3daed6e4fe              venture   
114947  /funding-round/35f09d0794651719b02bbfd859ba9ff5                 seed   
114948  /funding-round/af942869878d2cd788ef5189b435ebc4                grant   

       funding_round_code   funded_at  raised_amount_usd  
77                    NaN  12-08-2014                NaN  
729                     A  06-07-2015          5000000.0  
2670                  NaN  29-09-2014           100000.0  
3166                  NaN  12-10-2014                NaN  
3291                  NaN  15-10-2014           100000.0  
...                   ...         ...                ...  
110545                NaN  28-10-2014            30000.0  
113839                NaN  17-07-2010                NaN  
114946                  A  01-08-2014                NaN  
114947                NaN  01-01-2015            18192.0  
114948                NaN  01-10-2013            14851.0  

[74 rows x 6 columns]

In [31]:
# now it looks like all the columns are showing the values
# now let's check again the number of permalinks in rounds file

len(rounds['company_permalink'].unique())

66368

it works well. Now in both rounds and companies the number of permalinks are same.

Now let's check whther there is any missing permalink in companies file

In [33]:
companies.loc[~companies['permalink'].isin(rounds['company_permalink']),:]

permalink  \
43                               /organization/10â°north   
426                  /organization/51wofang-无忧我房   
1506                          /organization/adslinked™   
1775   /organization/aesthetic-everythingâ®-social-ne...   
1834              /organization/affluent-attachã©-club-2   
...                                                  ...   
63833                    /organization/whodat’s-spaces   
65778                 /organization/zengame-ç¦
游科暀   
66365                               /organization/ãeron   
66366                             /organization/ã”asys-2   
66367  /organization/ä°novatiff-reklam-ve-tanä±tä±m-h...   

                                                name  \
43                                         10°North   
426                            51wofang 无忧我房   
1506                                    AdsLinked™   
1775           Aesthetic Everything® Social Network   
1834                          Affluent Attaché Club   
...                                              ...   
63833                              Whodat’s Spaces   
65778                           ZenGame ç¦
游科技   
66365                                         ÁERON   
66366                                         Ôasys   
66367  İnovatiff Reklam ve Tanıtım Hizmetleri Tic   

                          homepage_url  \
43                                 NaN   
426            http://www.51wofang.com   
1506          http://www.adslinked.com   
1775   http://aestheticeverything.com/   
1834   http://www.affluentattache.com/   
...                                ...   
63833                              NaN   
65778          http://www.zen-game.com   
66365             http://www.aeron.hu/   
66366             http://www.oasys.io/   
66367             http://inovatiff.com   

                                           category_list     status  \
43                                               Fashion  operating   
426                                                  NaN     closed   
1506                                Advertising|Internet  operating   
1775                                    Public Relations  operating   
1834                                         Hospitality  operating   
...                                                  ...        ...   
63833                                               Apps  operating   
65778                Internet|Mobile Games|Online Gaming     closed   
66365                                                NaN  operating   
66366  Consumer Electronics|Internet of Things|Teleco...  operating   
66367                 Consumer Goods|E-Commerce|Internet  operating   

      country_code state_code       region            city  founded_at  
43             CAN         ON      Toronto     Mississauga  05-01-2013  
426            NaN        NaN          NaN             NaN  01-01-2014  
1506           NaN        NaN          NaN             NaN         NaN  
1775           USA         CA  Los Angeles  Marina Del Rey  15-06-2009  
1834           USA         CA  Los Angeles   Beverly Hills         NaN  
...            ...        ...          ...             ...         ...  
63833          NaN        NaN          NaN             NaN         NaN  
65778          NaN        NaN          NaN             NaN  17-07-2010  
66365          NaN        NaN          NaN             NaN  01-01-2011  
66366          USA         CA  SF Bay Area   San Francisco  01-01-2014  
66367          NaN        NaN          NaN             NaN         NaN  

[68 rows x 10 columns]

Thus company file also have some special characters. Let's take care of them first.

In [34]:
companies['permalink'] = companies['permalink'].str.encode('utf-8').str.decode('ascii', 'ignore')

Now let's check permalinks presents in companies data frame but not in round dataframe

In [35]:
companies.loc[~companies['permalink'].isin(rounds['company_permalink']),:]

Empty DataFrame
Columns: [permalink, name, homepage_url, category_list, status, country_code, state_code, region, city, founded_at]
Index: []

The encoding problem now solved. Let's save the file & go to next step

In [36]:
rounds.to_csv('round_final.csv', sep =',', index = False)
companies.to_csv('companies_final.csv', sep = '\t', index = False)